src/sbvr-api/permissions.ts

```diff
@@ -0,0 +1,1058 @@
+	if (apiKey == null || req.apiKey != null) {
+		return
+	}
+	return getApiKeyPermissions(apiKey)
+	.catch((err) => {
+		console.warn('Error with API key:', err)
+		// Ignore errors getting the api key and just use an empty permissions object
+		return []
+	})
+	.then((permissions) => {
+		req.apiKey = {
+			key: apiKey,
+			permissions: permissions
+		}
+	})
+})
+
+export const customAuthorizationMiddleware = (expectedScheme = 'Bearer') => {
+	if (expectedScheme == null) {
+		expectedScheme = 'Bearer';
+	}
+	expectedScheme = expectedScheme.toLowerCase()
+	return (req: express.Request, _res?: express.Response, next?: express.NextFunction): Promise<void> =>
+		Promise.try(() => {
+			const auth = req.header('Authorization')
+			if (!auth) {
+				return
+			}
+
+			const parts = auth.split(' ')
+			if (parts.length !== 2) {
+				return
+			}
+
+			const [ scheme, apiKey ] = parts
+			if (scheme.toLowerCase() !== expectedScheme) {
+				return
+			}
+
+			return checkApiKey(req, apiKey)
+		}).then(() => {
+			next && next()
+		})
+}
+
+// A default bearer middleware for convenience
+export const authorizationMiddleware = customAuthorizationMiddleware()
+
+export const customApiKeyMiddleware = (paramName = 'apikey') => {
+	if (paramName == null) {
+		paramName = 'apikey';
+	  }
+	return (req: HookReq | express.Request, _res?: express.Response, next?: express.NextFunction): Promise<void> => {
+		const apiKey = req.params[paramName] != null
+			? req.params[paramName]
+			: (
+				req.body[paramName] != null
+				? req.body[paramName]
+				: req.query[paramName]
+			)
+		return checkApiKey(req, apiKey)
+		.then(() => {
+			next && next()
+		})
+	}
+}
+
+// A default api key middleware for convenience
+export const apiKeyMiddleware = customApiKeyMiddleware()
+
+export const checkPermissions = (req: PermissionReq, actionList: PermissionCheck, resourceName?: string, vocabulary?: string) =>
+	getReqPermissions(req)
+	.then((permissionsLookup) =>
+		$checkPermissions(permissionsLookup, actionList, vocabulary, resourceName)
+	)
+
+export const checkPermissionsMiddleware = (action: PermissionCheck): express.RequestHandler =>
+	(req, res, next) =>
+		checkPermissions(req, action)
+		.then((allowed) => {
+			switch (allowed) {
+				case false:
+					res.sendStatus(401)
+					return
+				case true:
+					next()
+					return
+				default:
+					throw new Error('checkPermissionsMiddleware returned a conditional permission')
+			}
+		})
+		.catch((err) => {
+			sbvrUtils.api.Auth.logger.error('Error checking permissions', err, err.stack)
+			res.sendStatus(503)
+		})
+
+const getGuestPermissions = memoize(
+	() =>
+		// Get guest user
+		sbvrUtils.api.Auth.get({
+			resource: 'user',
+			passthrough: {
+				req: rootRead
+			},
+			options: {
+				$select: 'id',
+				$filter: {
+					username: 'guest'
+				}
+			}
+		})
+		.then((result: AnyObject[]) => {
+			if (result.length === 0) {
+				throw new Error('No guest permissions')
```

Comment:
I know that would be semantically different, but I think it would be better to just return an empty array.

Reply:
It is just surprising behaviour IMO, adding an extra special case that needs to be handled when it could be very well represented by the return value.

The function is called `getPermissions` IIRC so I'd expect it to return a set of permissions, including returning an empty set which unequivocally means the same thing as "no permissions". It should be up to the caller to decide whether that signifies an error condition.

Plus, I don't see why I, the caller, should be prepared to handle the "no permissions" case differently. I would only think to potentially catch errors like "the given user is not valid", or something similar.

Reply:
@dfunckt that makes sense and actually this case is the one that there is no guest user and the error message is wrong/confusing.  I've updated the error message to say user rather than permissions which should also make it clearer when anyone hits the error

Reply:
Haha, I completely missed that :) I sort of read `perms.length == 0` all along!